### PR TITLE
#24 Streamlit Dashboard Beta

### PR DESCRIPTION
**Pull request by anitamegarry**
Code to run the Streamlit dashboard. 

Current functionality:

- Users can create an account
- Users can log in
- Users can track new products by entering a URL and a notification price
- Users can view all products they are tracking: current price, original price, name of product, graph showing change in price over time, URL
- Users can log out

### CODE REVIEW

File: streamlit_login.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+            cursor.close()
+            conn.commit()
+            conn.close()
+            return get_subscription_id(user_id, product_id)
+    except Exception as e:
+        st.error(f"Error inserting into the database (subscription): {e}")
+        return None
+
+
+def insert_initial_price(price, product_id) -> None:
```
Comment (jk-george) [inline streamlit_login.py:339]: Insert functions have a similar "DRY" related issue. May want to consider slimming the code down.

**Review by jk-george — APPROVED** [review]
Very good work on the dashboard!

Just a few thoughts on "DRY" code.

File: dashboard.py [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 import altair as alt
 import streamlit as st
 import requests
 import psycopg2
 from bs4 import BeautifulSoup
-from streamlit_graphs import get_connection, get_cursor
+from database_connection import get_connection, get_cursor
+from streamlit_option_menu import option_menu
 
 headerSection = st.container()
```
Comment (indiah444) [inline dashboard.py:14]: Very small detail, but you may want to change this to snake case to match over variable names.

File: database_connection.py [hunk 3]
```diff
@@ -13,7 +13,7 @@ def get_connection():
     try:
         connection = psycopg2.connect(
             dbname=os.environ["DB_NAME"],
-            user=os.environ["DB_USERNAME"],
+            user=os.environ["DB_USER"],
```
Comment (indiah444) [inline database_connection.py:16]: Thank you for making this consistent!

File: dashboard.py [hunk 4]
```diff
@@ -374,27 +348,38 @@ def stop_tracking_product(user_id, product_id):
 def show_main_page():
     """Displays the main page on the dashboard"""
     with mainSection:
+        with st.sidebar:
+            page = option_menu(
+                menu_title="Menu", options=["About", "Current products", "Track new products", "Unsubscribe from product tracking"])
```
Comment (indiah444) [inline dashboard.py:353]: I think having this sidebar is such a good idea, it makes the dashboard a lot more organised!

**Review by indiah444 — APPROVED** [review]
This looks great, thank you for your hard work on this dashboard!

File: streamlit_login.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+    try:
+        conn = get_connection()
+        cursor = get_cursor(conn)
+        query = "SELECT user_id FROM users WHERE email_address = %s"
+        cursor.execute(query, (email,))
+        result = cursor.fetchone()
+        return result[0] if result else None
+    except psycopg2.Error as e:
+        print(f"Database error: {e}")
+        return None
```
[the remark this answers is not in the document]
Comment (anitamegarry) [reply streamlit_login.py]: Noted. Will refactor when all key functionality has been added

**Review by indiah444 — APPROVED** [review]
Thank you for making these changes, it all looks great!

**Pull request merged**